gradle-consistent-versions-idea-plugin/src/main/java/com/palantir/gradle/versions/intellij/GradleCacheExplorer.java

```diff
@@ -0,0 +1,175 @@
+            } catch (RuntimeException e) {
+                if (e.getCause() instanceof InterruptedException) {
+                    log.debug("Operation was cancelled", e);
+                    return Collections.emptySet();
+                }
+                throw e;
+            }
+        });
+    }
+
+    final boolean isValidResourceUrl(String url) {
+        return projectUrls.stream().anyMatch(url::startsWith) && (url.endsWith(".pom") || url.endsWith(".jar"));
+    }
+
+    final Stream<String> extractStringsFromBinFile(Path binFile) {
+        Set<String> result = new HashSet<>();
+        try (BufferedInputStream bis = new BufferedInputStream(Files.newInputStream(binFile))) {
+            StringBuilder currentString = new StringBuilder();
+            int byteValue;
+
+            while ((byteValue = bis.read()) != -1) {
+                char charValue = (char) byteValue;
+                if (!Character.isISOControl(charValue)) {
+                    currentString.append(charValue);
+                    continue;
+                }
+                if (!currentString.isEmpty()) {
+                    result.add(currentString.toString());
+                    currentString.setLength(0);
+                }
+            }
+
+            if (!currentString.isEmpty()) {
+                result.add(currentString.toString());
+            }
+        } catch (IOException e) {
+            log.error("Failed to extract strings from bin file", e);
+        }
+        return result.stream();
+    }
+
+    /**
+     * Extracts the group and artifact identifiers from a given URL.
+     *
+     * <p>The method removes the base project URL from the input if it matches any in a predefined list,
+     * then converts the remaining path to the format "group:artifact".
+     *
+     * <p>Example: For the URL "http://example.com/org/example/project/1.0/project-1.0.jar"
+     * and "http://example.com/" in the list of projectUrls, it returns "org.example.project:project".
+     *
+     * @param url the URL to process
+     * @return an {@link Optional} containing a string in the format "group:artifact" if extraction is successful,
+     *         or {@link Optional#empty()} if no matching project URL is found or the URL does not have the expected structure.
+     */
+    Optional<String> extractGroupAndArtifactFromUrl(String url) {
+        return projectUrls.stream().filter(url::startsWith).findFirst().flatMap(projectUrl -> {
+            String finalUrl = url.substring(projectUrl.length());
+            int lastSlashIndex = finalUrl.lastIndexOf('/');
+            int secondLastSlashIndex = finalUrl.lastIndexOf('/', lastSlashIndex - 1);
+
+            if (lastSlashIndex == -1 || secondLastSlashIndex == -1) {
+                return Optional.empty();
+            }
+
+            finalUrl = finalUrl.substring(0, secondLastSlashIndex).replace('/', '.');
+            finalUrl = finalUrl.replaceFirst("\\.([^.]*)$", ":$1");
```

Comment:
I'm still struggling to parse this a little. A bit more verbose, but could we do:

```java
int thirdLastSlashIndex = ...

String group = mavenLayout.substring(0, thirdLastSlashIndex).replace('/', '.');
String artifact = mavenLayout.substring(thirdLastSlashIndex, secondLastSlashIndex);
```